Northstar.CustomServers/mod/scripts/vscripts/weapons/sh_phase_shift.gnut

```diff
@@ -0,0 +1,763 @@
+	PrecacheParticleSystem( SHIFTER_DISAPPEAR_FX_TITAN )
+	PrecacheParticleSystem( SHIFTER_PERSONAL_FX )
+	PrecacheParticleSystem( SHIFTER_SCREEN_FX )
+	PrecacheParticleSystem( SHIFTER_SCREEN_FX_START )
+
+	MpAbilityShifterWeapon_Init()
+
+	RegisterSignal( "StartPhaseShift" )
+	RegisterSignal( "StopPhaseShift" )
+	RegisterSignal( "ForceStopPhaseShift" )
+}
+
+int function PhaseShift( entity ent, float warmupTime, float duration )
+{
+	if ( !IsAlive( ent ) )
+		return 0
+
+	if ( ent.IsPhaseShifted() )
+		return 0
+
+	// PROTO: this should be better eventually
+	// this is to prevent weirdness if you're already in a context action
+	if ( ent.IsPlayer() )
+	{
+		entity proxy = ent.GetFirstPersonProxy()
+		if ( ( ent.ContextAction_IsActive() || (proxy != null && proxy.Anim_IsActive()) )
+			&& !ent.IsZiplining()
+			&& !ent.ContextAction_IsLeeching()
+			&& ent.GetTitanSoulBeingRodeoed() == null )
+		{
+			return 0
+		}
+
+		ent.PhaseShiftBegin( warmupTime, duration )
+
+		#if SERVER
+			thread ServerPhaseShiftPlayerThread( ent, warmupTime, duration )
+		#else
+			thread ClientPhaseShiftWarmupThread( ent, warmupTime )
+		#endif // #if SERVER
+
+		return 1
+	}
+	else
+	{
+		ent.PhaseShiftBegin( warmupTime, duration )
+
+		#if SERVER
+			thread ServerPhaseShiftPlayerThread( ent, warmupTime, duration )
+		#endif // #if SERVER
+
+		return 1
+	}
+
+	unreachable
+}
+
+#if SERVER
+void function ServerPhaseShiftPlayerThread( entity ent, float warmupTime, float shiftTime )
+{
+	ent.EndSignal( "OnDeath" )
+	ent.EndSignal( "ForceStopPhaseShift" )
+
+	bool keepAnimActive = false
+
+	if ( ent.IsPlayer() )
+	{
+		ent.EndSignal( "player_embarks_titan" )
+
+		Leech_Disallow( ent )
+	//	EmitSoundOnEntityExceptToPlayer( ent, ent, SHIFTER_PRE_SOUND_3P )
+
+		entity proxy = ent.GetFirstPersonProxy()
+		if ( proxy != null )
+			proxy.SetForceVisibleInPhaseShift( true )
+
+		if ( Rodeo_IsAttached( ent ) )
+		{
+			if ( ent.GetParent() != null && ent.GetParent().IsPhaseShifted() )
+				keepAnimActive = true
+			else
+				ent.Signal( "RodeoOver" )
+		}
+
+		if ( ent.IsTitan() )
+		{
+			UntrackAllToneMarks( ent )
+			entity titanSoul = ent.GetTitanSoul()
+			if ( titanSoul != null )
+			{
+				if ( titanSoul.soul.batteryContainer != null )
+				{
+					titanSoul.soul.batteryContainer.SetForceVisibleInPhaseShift( true )
+				}
+			}
+		}
+
+		Battery_StopFXAndHideIconForPlayer( ent )
+	}
+	else
+	{
+		if ( ent.IsTitan() )
+			UntrackAllToneMarks( ent )
+		ent.EnableNPCFlag( NPC_IGNORE_ALL )
+		EmitSoundOnEntity( ent, SHIFTER_PRE_SOUND_3P )
+	}
+
+	//Why we need this one
+	//bool oldNoTarget = ent.GetNoTarget()
+
+	OnThreadEnd(
+		function() : ( ent, oldNoTarget )
+		{
+			if ( IsValid( ent ) )
+			{
+				ent.Signal( "StopPhaseShift" )
+				ent.Solid()
+				ent.SetNoTarget( false )
+				//Why we need "oldNoTarget",it'll case Ronin Bug;Common in Frontier Defense
+				//Npc will ignore the player and other team player in the this titan)
+				//ent.SetNoTarget( oldNoTarget )

```

Comment:
This probably existed because of other scripts setting players to not be targeted. We technically want to return back to the state that the player was in, as opposed to just setting them to be targetable. However because you can start another phase shift before you are back to that state, the second phase shift will always set you to be non-targetable after it finishes

Reply:
After I use  `false` instead of `oldNoTarget`,everything is fine.(except the value of `oldNoTarget`)
Even if I phasing two times in quick succession again.

Reply:
What happens if your `ent.SetNoTarget` was `true` because another script set it to be `true` earlier? It looks like it would become `false`, which could cause problems for other scripts that might expect it to be `true`

Reply:
didn't think it'll case other problems
`oldNoTarget` only use in the `OnThreadEnd`,not in any other place

Test Log all good like normal:
Start No.1 Phase Shift: false
Start No.1 No target before: false
Start No.1 No target after: true
End No.1 Phase Shift: false
Start No.2 Phase Shift: false
Start No.2 No target before: false
Start No.2 No target after: true
End No.2 Phase Shift: false

The normal process is Call the function `CodeCallback_EnterPhaseShift` or `PhaseShift` -> `bool oldNoTarget = ent.GetNoTarget()` -> oldNoTarget=false -> Start Phase Shift -> End Phase Shift ->Call the function `CodeCallback_ExitPhaseShift` or `CancelPhaseShift` -> Trigger `ent.Signal( "ForceStopPhaseShift" )` ->Trigger `ent.EndSignal( "ForceStopPhaseShift" )` -> Trigger `OnThreadEnd` -> ent.SetNoTarget( oldNoTarget ) (PS:This value is `false`)

I guess because of phasing two times in quick succession
In the first time,the `CodeCallback_ExitPhaseShift` isn't triggered.
And the Signal `ForceStopPhaseShift` also not triggered.
So,Ronin still in the NoTarget state.
`OnThreadEnd` will delayed trigger
After the second time enter the Phase Shift
When the game try run `bool oldNoTarget = ent.GetNoTarget()`,it'll return `true`
After all of these,when the second time Phase Shift ended;
Game will enter the `OnThreadEnd`,it'll trigger `ent.SetNoTarget( oldNoTarget )`.But the `oldNoTarget` is `true` :D

Reply:
What about the following case:

Player uses the console command `notarget`. `player.GetNoTarget()` -> `true`
Player uses phase shift. `player.GetNoTarget()` -> `true`
Player exists phase shift. `player.GetNoTarget()` -> `false`

The player would lose the effect of the command `notarget`

Reply:
You mean add a parameter to the function like this?
int function PhaseShift( entity ent, float warmupTime, float duration ,bool shouldKeepNoTarget = false )
void function ServerPhaseShiftPlayerThread( entity ent, float warmupTime, float shiftTime ,bool shouldKeepNoTarget = false )
...
OnThreadEnd(
		function() : ( ent, oldNoTarget )
		{
			if ( IsValid( ent ) )
			{
				ent.Signal( "StopPhaseShift" )
				ent.Solid()
				ent.SetNoTarget( shouldKeepNoTarget )